Inform users about server's username rules on invalid username error

ctrlaltca opened the pull request:
This PR aims at resolving #1173.
When a register or login command fails because the username is not accepted as valid, the server will include in the error response the rules it's actually using to validate the username.
The client will unpack these rules and try to explain them to the user:
![popup](https://cloud.githubusercontent.com/assets/1631111/8436466/d1bfaba0-1f58-11e5-9234-aec1d2580c10.png)

The proposed examples in #1173 were more specific about the error, but unfortunately we're parsing the username using a regexp, so we can't really tell the user why the username failed.
This change is fully backwards compatible on the client; if the server is not updated, the old "hardcoded" message will be shown to the user.

ctrlaltca commented:
I accept suggestions on the wording; will will need to end up in translations, too.

ctrlaltca commented on cockatrice/src/window_main.cpp:
```diff
@@ -79,7 +83,7 @@ void MainWindow::processConnectionClosedEvent(const Event_ConnectionClosed &even
         }
         case Event_ConnectionClosed::SERVER_SHUTDOWN: reasonStr = tr("Scheduled server shutdown."); break;
-        case Event_ConnectionClosed::USERNAMEINVALID: reasonStr = tr("Invalid username.\nYou may only use A-Z, a-z, 0-9, _, ., and - in your username."); break;
+        case Event_ConnectionClosed::USERNAMEINVALID: reasonStr = tr("Invalid username."); break;
```
Note: this code execution path is unused.

ZeldaZach replied:
If unused, then trash it :smile:

Daenyth replied:
I think it needs to be there for the compiler

On Wed, Jul 1, 2015 at 6:45 PM Zach <notifications@github.com> wrote:

> In cockatrice/src/window_main.cpp
> <https://github.com/Cockatrice/Cockatrice/pull/1180#discussion_r33732908>:
>
> > @@ -79,7 +83,7 @@ void MainWindow::processConnectionClosedEvent(const Event_ConnectionClosed &even
> >              break;
> >          }
> >          case Event_ConnectionClosed::SERVER_SHUTDOWN: reasonStr = tr("Scheduled server shutdown."); break;
> > -        case Event_ConnectionClosed::USERNAMEINVALID: reasonStr = tr("Invalid username.\nYou may only use A-Z, a-z, 0-9, _, ., and - in your username."); break;
> > +        case Event_ConnectionClosed::USERNAMEINVALID: reasonStr = tr("Invalid username."); break;
>
> If unused, then trash it [image: :smile:]
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Cockatrice/Cockatrice/pull/1180/files#r33732908>.
>

woogerboy21 commented:
So are you actually constructing the error string and then passing that string to the client to display in the popup box?

ctrlaltca commented:
No, that won't permit translations. I'm passing the client a crappy string containing rules:
https://github.com/Cockatrice/Cockatrice/pull/1180/files#diff-e8469b96aeb5655c901ad95630a25753R130
The client then unrolls the settings string and builds up the fully-translated message.

poixen commented:
If new rules are added to the server and ini, will they be sent too?

ctrlaltca commented:
Not until you add them to the server, tell the server to send them and instruct cockatrice to interpret what they mean and inform the user accordingly.

woogerboy21 commented:
@poixen : it does look like it reads the current variable values and displays them accordingly.  So if a server operator changes say the special character restrictions it picks that up. (Or am i reading that incorrectly @ctrlaltca )

ctrlaltca commented:
Yes, the current value of existing settings are shown, including the current valid punctuation characters.

woogerboy21 commented:
I dont see the check for "allowpunctuationprefix".

ctrlaltca commented:
It's not in my screenshot, but it's implemented here: https://github.com/Cockatrice/Cockatrice/pull/1180/files#diff-06fec504bab8f0aa370636aea5d74fdfR360

poixen commented:
Really need to add a post punctuation check too. So ctrlaltca, for new rules, we just update the server and ini, then the message we added in the server will be sent to the client?

ctrlaltca commented:
You can't send directly the popup message from the server, because the client won't be able to translate it. That's why i had to implement a rudimentary the rule parsing in cockatrice.

Daenyth commented on cockatrice/src/window_main.cpp:
```diff
@@ -328,6 +335,36 @@ void MainWindow::loginError(Response::ResponseCode r, QString reasonStr, quint32
     actConnect();
 }
 
+void MainWindow::extractInvalidUsernameMessage(QString & in, QString & out)
```
Is there a recommended practice for this?

Using `void` signature with an out parameter seems wrong to me. Why not return the out value normally?

ctrlaltca commented:
@Daenyth i've changed the method signature to be a more common method receiving a QString parameter and return a QString value.
@woogerboy21 your comment about "allowpunctuationprefix" was correct, the check was inverted. Fixed this one, too.

woogerboy21 commented:
This works, but would it be to much trouble to add a log entry about the server configuration file being reloaded when te sighup is received?

woogerboy21 commented:
Here is what is shown if the username does not meet the character limits.
![capture](https://cloud.githubusercontent.com/assets/1361287/8505474/42a325ec-21b4-11e5-8562-f4fa68125d1e.JPG)

ZeldaZach commented:
I did testing on what @woogerboy21 and it appears he was using an outdated client with a potentially outdated server. When I ran a server on the newest everything, the errors appeared appropriately.

In addition, @ctrlaltca, I've made a small change to how the rules are shown:
<img width="428" alt="screenshot 2015-07-03 23 55 46" src="https://cloud.githubusercontent.com/assets/7460172/8506381/1bffa9ee-21df-11e5-9473-c995b727869c.png">

As you can see, all rules will be shown and if you're not allowed to do something, it will show the "NOT" word in the message.

You can find my change here: https://github.com/ZeldaZach/Cockatrice/commit/1d210e57bd3707f01f54be512b042b45fc15d508

Finally, if possible, can we find a way to align the bullet points to the left to use less room? This isn't a major issue, but if possible I'd really appreciate it.

Hope I was of some help :)
